nntrainer/opencl/opencl_program.cpp

```diff
@@ -41,6 +43,105 @@ bool Program::BuildProgram(cl_device_id device_id,
     return false;
   }
 
+  // saving kernel binary
+  if (!binaryCreated)
+    return GetProgramInfo(device_id);
+
+  return true;
+}
+
+/**
+ * @brief Utility to get program info and save kernel binaries
+ *
+ * @param device_id OpenCL device id
+ * @return true if successful or false otherwise
+ */
+bool Program::GetProgramInfo(cl_device_id device_id) {
+  // since only GPU is being used
```

Comment:
You are writing the whole function as a C code without using convinient (making the code shorter, simpler, and more readable) C++ syntax. The pointers in this funcion are supposed to be freed on exit, try smart pointers (e.g., std::unique_ptr).